lib/MusicBrainz/Server/Entity/Role/Linkable.pm

```diff
@@ -75,13 +75,20 @@ sub relationships_by_link_type_names
     } $self->all_relationships ];
 }
 
+my $_serialize_relationships = 1;
+
 around TO_JSON => sub {
     my ($orig, $self) = @_;
 
     my $json = $self->$orig;
 
-    if ($self->has_loaded_relationships) {
-        $json->{relationships} = [map { $_->TO_JSON } $self->all_relationships];
+    if ($_serialize_relationships && $self->has_loaded_relationships) {
+        $json->{relationships} = [map {
+            $_serialize_relationships = 0;
+            my $result = $_->TO_JSON;
+            $_serialize_relationships = 1;
+            $result;
+        } $self->all_relationships];
```

Comment:
You can have cyclic references via relationships. Like Artist1.relationships @> Work1 and Work1.relationships @> Artist1. Calling TO_JSON on Artist1 would serialize the Work1 relationship, which would serialize Work1, then Work1.relationships, then Artist1 again, leading to infinite recursion.

It wasn't a problem before, because the serialization was lazy -- by the time we got around to serializing Work1, Artist1 would have already been added to `$linked_entities`, so it would be skipped on the occurrence.

The commit titled "Further avoid infinite recursion in add_linked_entity" (I won't link to it directly because I'll likely break the commit ID later) doesn't fix this because we don't output relationship targets via `$linked_entities`, we just output them directly.

Reply:
After looking into this again, I can't find where I triggered this kind of infinite recursion. I did somewhere, but...

I'll keep this check just in case, but take the explanation above with a grain of salt. It's safer to keep around until we move relationship targets to `$linked_entities`.